program/src/processor/inactivate_stake.rs

```diff
@@ -13,8 +21,90 @@ use crate::instruction::accounts::{Context, InactivateStakeAccounts};
 /// 0. `[w]` Stake config account
 /// 1. `[w]` Validator stake account
 pub fn process_inactivate_stake(
-    _program_id: &Pubkey,
-    _ctx: Context<InactivateStakeAccounts>,
+    program_id: &Pubkey,
+    ctx: Context<InactivateStakeAccounts>,
 ) -> ProgramResult {
-    Ok(())
+    // Account validation.
+
+    // config
+    // - owner must be the stake program
+    // - must be initialized
+
+    require!(
+        ctx.accounts.config.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "config"
+    );
+
+    let data = &mut ctx.accounts.config.try_borrow_mut_data()?;
+    let config = bytemuck::try_from_bytes_mut::<Config>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        config.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "config",
+    );
+
+    // stake
+    // - owner must be the stake program
+    // - must be initialized
+    // - must have the correct derivation
+
+    require!(
+        ctx.accounts.stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "stake"
+    );
+
+    let data = &mut ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    // validates that the stake account corresponds to the received
+    // config account
+    let (derivation, _) = find_stake_pda(&stake.validator, ctx.accounts.config.key, program_id);
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // Inactivates the stake if elegible.
+
+    if let Some(timestamp) = stake.deactivation_timestamp {
+        let current = Clock::get()?
+            .unix_timestamp
+            .saturating_sub(config.cooldown_time_seconds) as u64;
+
+        require!(
+            current >= timestamp.get(),
+            StakeError::ActiveDeactivationCooldown,
+            "{} second(s) remaining for deactivation",
+            timestamp.get().saturating_sub(current),
+        );
+
+        msg!("Deactivating {} token(s)", stake.deactivating_amount);
+
+        // moves deactivating amount to inactive
+        stake.amount = stake.amount.saturating_sub(stake.deactivating_amount);
```

Comment:
This is the paranoid part of me, but how about doing a `checked_sub` instead? This should never fail, right?

Reply:
I will change it to `checked_sub` – and yes, this should never fail since the `deactivating_amount` will always be smaller or equal to the amount.